File: poly_apps/laravel_main/app/Services/AiGateway/AiChat.php
<?php

namespace App\Services\AiGateway;

use Illuminate\Support\Facades\Http;

/**
 * Unified AI chat — the PHP port of pycore's pyctl.ai.ai_chat.chat_once.
 *
 * Sends one real chat turn to a single provider and returns the unified
 * contract the UI depends on:
 *   { success, provider, model, nickname, text, latency_ms, error, retry_after_s }
 *
 * Keys are loaded with the SAME indexed precedence as the probe (AiSecretLoader
 * via AiProviderRegistry). Most providers go through the OpenAI-compatible
 * client; gemini / anthropic / cloudflare speak their own dialect and are built
 * directly here on the Http facade.
 */
class AiChat
{
    private const VALID_ROLES = ['system', 'user', 'assistant'];

    /**
     * @param array<int, array<string, mixed>> $messages
     */
    public static function chatOnce(string $provider, array $messages, ?string $model = null, string $source = ''): array
    {
        $provider = strtolower(trim($provider));
        $requestedModel = trim((string) $model) !== '' ? trim((string) $model) : null;
        $out = self::blankResult($provider, $requestedModel ?? '');

        if (!AiProviderRegistry::exists($provider)) {
            $out['error'] = "Unknown provider: '{$provider}'";
            return $out;
        }

        $msgs = self::normalizeMessages($messages);
        if (empty($msgs)) {
            $out['error'] = 'No message provided';
            return $out;
        }

        // Image-only providers (pollinations / imagen / azure / bedrock / vertex)
        // have no chat backend.
        if (AiProviderRegistry::isImageOnly($provider)) {
            $out['error'] = "Provider '{$provider}' is image-only (no chat backend)";
            return $out;
        }

        if (!AiProviderRegistry::isConfigured($provider)) {
            $out['error'] = 'No API key configured';
            return $out;
        }

        $useModel = $requestedModel ?: AiProviderRegistry::defaultModel($provider);

        $rate = AiRateLimiter::checkRateLimit($provider, $useModel);
        if (!$rate['allowed']) {
            $out['error'] = $rate['message'];
            $out['retry_after_s'] = $rate['retry_after_s'];
            $out['model'] = $useModel;
            $out['nickname'] = self::nickname($provider, $useModel);
            AiUsageLog::record('text', $provider, $useModel, false, null, $source, $rate['message']);
            return $out;
        }

        // Multi-key rotation with PER-KEY cooldown + counters (AiKeyRotation, the
        // twin of pycore ai_key_rotation): pick the active (non-cooled) key; on an
        // AUTH/QUOTA error cool THAT key so it's SKIPPED next time, then rotate to
        // the next key; record every attempt per-key and gate on a per-key rate
        // budget (each key = its own account/quota).
        $keys = AiProviderRegistry::allSecrets($provider);
        if (empty($keys)) {
            $keys = [''];
        }
        [$rpm, $rpd] = AiRateLimiter::perKeyCaps($provider, $useModel);
        $n = count($keys);
        $start = microtime(true);
        for ($attempt = 0; $attempt < $n; $attempt++) {
            [$idx, $key] = AiKeyRotation::selectActive($provider, $keys);
            if (!AiKeyRotation::rateOk($provider, $idx, $rpm, $rpd)) {
                if ($attempt + 1 < $n) {
                    AiKeyRotation::markCooldown($provider, $idx, 30, 'per-key rate budget');
                    continue;
                }
                $out['error'] = 'per-key rate budget reached (all keys)';
                break;
            }
            $out['success'] = false;
            $out['error'] = null;
            try {
                self::dispatch($provider, $msgs, $requestedModel, $key, $out);
            } catch (\Throwable $e) {
                $out['error'] = $e->getMessage();
            }
            AiKeyRotation::record($provider, $idx, !empty($out['success']), $out['error'] ?? null);
            if (!empty($out['success'])) {
                break;
            }
            if (self::isAuthOrQuotaError($out['error'] ?? null)) {
                AiKeyRotation::markCooldown($provider, $idx, null, $out['error'] ?? null);
                if ($attempt + 1 < $n) {
                    continue;
                }
            }
            break;
        }

        if (empty($out['model'])) {
            $out['model'] = AiProviderRegistry::defaultModel($provider);
        }
        $out['nickname'] = self::nickname($provider, (string) $out['model']);
        $out['latency_ms'] = round((microtime(true) - $start) * 1000, 1);

        if (!empty($out['success'])) {
            AiRateLimiter::recordRequest($provider);
        }
        AiUsageLog::record('text', $provider, (string) $out['model'], (bool) $out['success'],
            $out['latency_ms'], $source, $out['error'] ?? null);
        return $out;
    }

    /**
     * Route one turn to the matching provider dialect, filling $out by reference.
     *
     * @param array<int, array{role:string, content:string}> $messages
     */
    private static function dispatch(string $provider, array $messages, ?string $model, string $key, array &$out): void
    {
        switch (AiProviderRegistry::client($provider)) {
            case 'gemini':
                self::chatGemini($provider, $messages, $model, $key, $out);
                return;
            case 'anthropic':
                self::chatAnthropic($provider, $messages, $model, $key, $out);
                return;
            case 'cloudflare':
                self::chatCloudflare($provider, $messages, $model, $key, $out);
                return;
            default:
                // compat + spark (the modern Spark endpoint is OpenAI-compatible).
                self::chatCompat($provider, $messages, $model, $key, $out);
        }
    }

    private static function chatCompat(string $provider, array $messages, ?string $model, string $key, array &$out): void
    {
        $useModel = $model ?: AiProviderRegistry::defaultModel($provider);
        $out['model'] = $useModel;
        $client = new OpenAiCompatClient(
            AiProviderRegistry::baseUrl($provider),
            $key,
            AiProviderRegistry::extraHeaders($provider)
        );
        $res = $client->chatCompletion($messages, $useModel);
        $out['text'] = $res['text'];
        $out['success'] = $res['success'];
        if (!$res['success']) {
            $out['error'] = $res['error'] ?: 'Empty response from provider';
        }
    }

    private static function chatGemini(string $provider, array $messages, ?string $model, string $key, array &$out): void
    {
        $useModel = $model ?: AiProviderRegistry::defaultModel($provider);
        $out['model'] = $useModel;
        $base = AiProviderRegistry::baseUrl($provider);
        $url = $base . '/models/' . rawurlencode($useModel) . ':generateContent';
        $response = Http::withHeaders(['Content-Type' => 'application/json'])
            ->timeout(90)
            ->post($url . '?key=' . urlencode($key), [
                'contents' => [[
                    'parts' => [['text' => self::messagesToPrompt($messages)]],
                ]],
            ]);

        if ($response->status() !== 200) {
            $out['error'] = 'HTTP ' . $response->status() . ': ' . mb_substr($response->body(), 0, 300);
            return;
        }
        $data = $response->json();
        $text = '';
        foreach ($data['candidates'][0]['content']['parts'] ?? [] as $part) {
            $text .= $part['text'] ?? '';
        }
        $out['text'] = $text;
        $out['success'] = $text !== '';
        if ($text === '') {
            $out['error'] = 'Empty response from provider';
        }
    }

    private static function chatAnthropic(string $provider, array $messages, ?string $model, string $key, array &$out): void
    {
        $useModel = $model ?: AiProviderRegistry::defaultModel($provider);
        $out['model'] = $useModel;

        $system = [];
        $turns = [];
        foreach ($messages as $m) {
            if ($m['role'] === 'system') {
                $system[] = $m['content'];
            } elseif (in_array($m['role'], ['user', 'assistant'], true)) {
                $turns[] = ['role' => $m['role'], 'content' => $m['content']];
            }
        }
        if (empty($turns)) {
            $turns = [['role' => 'user', 'content' => self::messagesToPrompt($messages)]];
        }
        $body = ['model' => $useModel, 'max_tokens' => 1024, 'messages' => $turns];
        if (!empty($system)) {
            $body['system'] = implode("\n", $system);
        }

        $response = Http::withHeaders([
            'x-api-key' => $key,
            'anthropic-version' => '2023-06-01',
            'Content-Type' => 'application/json',
        ])->timeout(90)->post(AiProviderRegistry::baseUrl($provider) . '/messages', $body);

        if ($response->status() !== 200) {
            $out['error'] = 'HTTP ' . $response->status() . ': ' . mb_substr($response->body(), 0, 300);
            return;
        }
        $data = $response->json();
        $text = '';
        foreach ($data['content'] ?? [] as $block) {
            if (($block['type'] ?? '') === 'text') {
                $text .= $block['text'] ?? '';
            }
        }
        $out['text'] = $text;
        $out['success'] = $text !== '';
        if ($text === '') {
            $out['error'] = 'Empty response from provider';
        }
    }

    private static function chatCloudflare(string $provider, array $messages, ?string $model, string $key, array &$out): void
    {
        $useModel = $model ?: AiProviderRegistry::defaultModel($provider);
        $out['model'] = $useModel;
        $accountId = AiProviderRegistry::extraSecret($provider);
        if ($accountId === '') {
            $out['error'] = 'CLOUDFLARE_ACCOUNT_ID not configured';
            return;
        }
        $url = AiProviderRegistry::baseUrl($provider) . '/accounts/' . $accountId . '/ai/run/' . $useModel;
        $response = Http::withHeaders([
            'Authorization' => 'Bearer ' . $key,
            'Content-Type' => 'application/json',
        ])->timeout(90)->post($url, ['messages' => $messages]);

        if ($response->status() !== 200) {
            $out['error'] = 'HTTP ' . $response->status() . ': ' . mb_substr($response->body(), 0, 300);
            return;
        }
        $data = $response->json();
        $text = (string) ($data['result']['response'] ?? '');
        $out['text'] = $text;
        $out['success'] = $text !== '';
        if ($text === '') {
            $out['error'] = 'Empty response from provider';
        }
    }

    // --- helpers ----------------------------------------------------------- //

    /** Error fragments meaning AUTH / QUOTA — these trigger a key rotation. */
    private const AUTH_QUOTA_MARKS = [
        'http 401', 'http 403', 'http 429',
        'rate limit', 'rate_limit', 'ratelimit',
        'quota', 'insufficient', 'overloaded', 'invalid api key', 'invalid_api_key',
    ];

    /**
     * True when an error is an AUTH/QUOTA failure (HTTP 401/403/429 or a quota /
     * rate-limit / invalid-key message) — the signal to try the NEXT key. Any
     * other error stops the rotation (it would fail identically on every key).
     */
    private static function isAuthOrQuotaError(?string $error): bool
    {
        $e = strtolower((string) $error);
        if ($e === '') {
            return false;
        }
        foreach (self::AUTH_QUOTA_MARKS as $mark) {
            if (str_contains($e, $mark)) {
                return true;
            }
        }
        return false;
    }

    public static function nickname(string $provider, string $model): string
    {
        $model = trim($model);
        return $model !== '' ? "{$provider}/{$model}" : $provider;
    }

    /**
     * @param array<int, array<string, mixed>> $messages
     * @return array<int, array{role:string, content:string}>
     */
    private static function normalizeMessages(array $messages): array
    {
        $out = [];
        foreach ($messages as $m) {
            if (!is_array($m) || !array_key_exists('content', $m) || $m['content'] === null) {
                continue;
            }
            $role = strtolower(trim((string) ($m['role'] ?? 'user')));
            if (!in_array($role, self::VALID_ROLES, true)) {
                $role = 'user';
            }
            $out[] = ['role' => $role, 'content' => (string) $m['content']];
        }
        return $out;
    }

    /** Flatten a message list into a transcript prompt (for Gemini). */
    private static function messagesToPrompt(array $messages): string
    {
        $parts = [];
        foreach ($messages as $m) {
            if ($m['role'] === 'system') {
                $parts[] = $m['content'];
            } elseif ($m['role'] === 'assistant') {
                $parts[] = 'Assistant: ' . $m['content'];
            } else {
                $parts[] = 'User: ' . $m['content'];
            }
        }
        return trim(implode("\n", $parts));
    }

    private static function blankResult(string $provider, string $model): array
    {
        return [
            'success' => false,
            'provider' => $provider,
            'model' => $model,
            'nickname' => self::nickname($provider, $model),
            'text' => '',
            'latency_ms' => null,
            'error' => null,
            'retry_after_s' => null,
        ];
    }
}

File: pycore/pyctl/ai/ai_chat.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Unified AI chat — minimal "does it actually answer?" confirmation.

Sibling of :mod:`pycore.pyctl.ai.ai_probe`. The probe only proves a key can
*list models*; this module sends a real chat message to a provider and returns
the assistant's reply, so the desktop UI can confirm a provider truly works by
clicking it and chatting.

It loads each provider's key with the SAME key-name precedence as ai_probe (so
chat uses the exact key the probe validated), instantiates the matching pyutils
client directly (never the shared singleton, to avoid a stale key), and runs one
blocking completion. Gemini has no multi-turn message API here, so the message
list is flattened into a transcript prompt.

Contract (UI depends on this shape):
    {
      "success": bool,
      "provider": str,
      "model": str,
      "nickname": str,                # provider/model display label
      "text": str,
      "latency_ms": number | None,
      "error": str | None,
      "retry_after_s": number | None  # local rate-limit backoff hint
    }
"""

import time
from functools import partial
from typing import Dict, Any, List, Optional

from pycore.pyfoundations.pybasecommon.color_print import ColorPrint
from pycore.pyfoundations.third_party.api import get_third_package_requests
from pycore.pyctl.ai.ai_keys import PROVIDERS, first_secret, default_model, OPENAI_COMPAT_PROVIDERS, is_configured
from pycore.pyctl.ai.ai_rate_limits import check_rate_limit, record_request, chat_nickname
from pycore.pyctl.ai.ai_compat_helpers import chat_openai_compat, chat_cloudflare, chat_spark
from pycore.pyctl.ai.ai_usage_log import record_usage

from pycore.pyutils.ai_cluster.openrouter.openrouter_client import OpenRouterClient
from pycore.pyutils.ai_cluster.gemini.gemini_client import GeminiClient
from pycore.pyutils.ai_cluster.deepseek.deepseek_client import get_deepseek_client
from pycore.pyutils.ai_cluster.groq.groq_client import GroqClient
from pycore.pyutils.ai_cluster.mistral.mistral_client import MistralClient
from pycore.pyutils.ai_cluster.cohere.cohere_client import CohereClient
from pycore.pyutils.ai_cluster.nvidia.nvidia_client import NVIDIAClient
from pycore.pyutils.ai_cluster.huggingface.hf_client import HuggingFaceClient
from pycore.pyutils.ai_cluster.cerebras.cerebras_client import CerebrasClient
from pycore.pyutils.ai_cluster.github.github_client import GitHubModelsClient
from pycore.pyutils.ai_cluster.zhipuai.zhipuai_client import ZhipuAIClient


_VALID_ROLES = ("system", "user", "assistant")


def _normalize_messages(messages: Optional[List[Dict[str, Any]]]) -> List[Dict[str, str]]:
    """Coerce arbitrary message dicts into clean {role, content} pairs."""
    out: List[Dict[str, str]] = []
    for m in messages or []:
        if not isinstance(m, dict):
            continue
        content = m.get("content")
        if content is None:
            continue
        role = str(m.get("role") or "user").strip().lower()
        if role not in _VALID_ROLES:
            role = "user"
        out.append({"role": role, "content": str(content)})
    return out


def _messages_to_prompt(messages: List[Dict[str, str]]) -> str:
    """Flatten a message list into a transcript prompt (for Gemini)."""
    parts: List[str] = []
    for m in messages:
        if m["role"] == "system":
            parts.append(m["content"])
        elif m["role"] == "assistant":
            parts.append(f"Assistant: {m['content']}")
        else:
            parts.append(f"User: {m['content']}")
    return "\n".join(parts).strip()


def _result(provider: str, model: str) -> Dict[str, Any]:
    return {
        "success": False,
        "provider": provider,
        "model": model,
        "nickname": chat_nickname(provider, model),
        "text": "",
        "latency_ms": None,
        "error": None,
        "retry_after_s": None,
    }


def _openrouter_first_model(key: str) -> Optional[str]:
    """Pick a currently-live OpenRouter model (prefer a free one) for no-model calls.

    OpenRouter's catalog changes, so a hardcoded default can 404. The UI always
    sends a probed model (fast path); this is only used when a caller omits one.
    """
    try:
        requests = get_third_package_requests()
        resp = requests.get(
            "https://openrouter.ai/api/v1/models",
            headers={"Authorization": f"Bearer {key}"},
            timeout=15,
        )
        resp.raise_for_status()
        data = resp.json().get("data", [])
        for m in data:
            mid = m.get("id", "")
            if mid == "openrouter/free":
                return mid
        for m in data:
            mid = m.get("id", "")
            if mid.endswith(":free"):
                return mid
        return data[0].get("id") if data else None
    except Exception:
        return None


def _chat_openrouter(messages, model, key, out):
    client = OpenRouterClient(api_key=key)
    if not model:
        model = _openrouter_first_model(key) or default_model("openrouter")
    out["model"] = model
    resp = client.chat_completion(messages=messages, model=model)
    if isinstance(resp, dict) and resp.get("error"):
        out["error"] = str(resp["error"])
        return out
    message = (resp.get("choices") or [{}])[0].get("message", {}) if isinstance(resp, dict) else {}
    out["text"] = client._extract_message_content(message)
    out["success"] = bool(out["text"])
    if not out["success"]:
        out["error"] = "Empty response from provider"
    return out


def _chat_gemini(messages, model, key, out):
    model = model or default_model("gemini")
    out["model"] = model
    client = GeminiClient(api_key=key, default_model=model)
    res = client.generate_content(prompt=_messages_to_prompt(messages), model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


def _chat_deepseek(messages, model, key, out):
    model = model or default_model("deepseek")
    out["model"] = model
    client = get_deepseek_client(api_key=key)
    res = client.chat_completion(messages=messages, model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


def _chat_openai(messages, model, key, out):
    """ChatGPT via the OpenAI chat-completions REST API (no SDK dependency)."""
    requests = get_third_package_requests()
    model = model or default_model("openai")
    out["model"] = model
    resp = requests.post(
        "https://api.openai.com/v1/chat/completions",
        headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"},
        json={"model": model, "messages": messages},
        timeout=60,
    )
    if resp.status_code != 200:
        out["error"] = f"HTTP {resp.status_code}: {resp.text[:300]}"
        return out
    data = resp.json()
    out["text"] = ((data.get("choices") or [{}])[0].get("message") or {}).get("content") or ""
    out["success"] = bool(out["text"])
    if not out["success"]:
        out["error"] = "Empty response from provider"
    return out


def _chat_anthropic(messages, model, key, out):
    """Claude via the Anthropic Messages REST API (system goes in its own field)."""
    requests = get_third_package_requests()
    model = model or default_model("anthropic")
    out["model"] = model
    system = "\n".join(m["content"] for m in messages if m["role"] == "system") or None
    turns = [m for m in messages if m["role"] in ("user", "assistant")]
    if not turns:
        turns = [{"role": "user", "content": _messages_to_prompt(messages)}]
    body = {"model": model, "max_tokens": 1024, "messages": turns}
    if system:
        body["system"] = system
    resp = requests.post(
        "https://api.anthropic.com/v1/messages",
        headers={
            "x-api-key": key,
            "anthropic-version": "2023-06-01",
            "Content-Type": "application/json",
        },
        json=body,
        timeout=60,
    )
    if resp.status_code != 200:
        out["error"] = f"HTTP {resp.status_code}: {resp.text[:300]}"
        return out
    data = resp.json()
    out["text"] = "".join(
        b.get("text", "") for b in (data.get("content") or []) if b.get("type") == "text"
    )
    out["success"] = bool(out["text"])
    if not out["success"]:
        out["error"] = "Empty response from provider"
    return out


def _chat_groq(messages, model, key, out):
    """Groq via its OpenAI-compatible API."""
    model = model or default_model("groq")
    out["model"] = model
    client = GroqClient(api_key=key, default_model=model)
    res = client.chat_completion(messages=messages, model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


def _chat_mistral(messages, model, key, out):
    """Mistral AI via its chat completions API."""
    model = model or default_model("mistral")
    out["model"] = model
    client = MistralClient(api_key=key, default_model=model)
    res = client.chat_completion(messages=messages, model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


def _chat_cohere(messages, model, key, out):
    """Cohere via its chat API."""
    model = model or default_model("cohere")
    out["model"] = model
    client = CohereClient(api_key=key, default_model=model)
    res = client.chat_completion(messages=messages, model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


def _chat_nvidia(messages, model, key, out):
    """NVIDIA NIM via its OpenAI-compatible API."""
    model = model or default_model("nvidia")
    out["model"] = model
    client = NVIDIAClient(api_key=key, default_model=model)
    res = client.chat_completion(messages=messages, model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


def _chat_huggingface(messages, model, key, out):
    """HuggingFace Inference API via its client."""
    model = model or default_model("huggingface")
    out["model"] = model
    client = HuggingFaceClient(api_key=key, default_model=model)
    res = client.chat_completion(messages=messages, model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


def _chat_cerebras(messages, model, key, out):
    model = model or default_model("cerebras")
    out["model"] = model
    client = CerebrasClient(api_key=key, default_model=model)
    res = client.chat_completion(messages=messages, model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


def _chat_github(messages, model, key, out):
    model = model or default_model("github")
    out["model"] = model
    client = GitHubModelsClient(api_key=key, default_model=model)
    res = client.chat_completion(messages=messages, model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


def _chat_zhipuai(messages, model, key, out):
    """Zhipu AI (BigModel.cn) via its OpenAI-compatible API."""
    model = model or default_model("zhipuai")
    out["model"] = model
    client = ZhipuAIClient(api_key=key, default_model=model)
    res = client.chat_completion(messages=messages, model=model)
    if res.get("success"):
        out["text"] = res.get("text", "")
        out["success"] = True
    else:
        out["error"] = res.get("error") or "Empty response from provider"
    return out


_DISPATCH = {
    "openrouter": _chat_openrouter,
    "gemini": _chat_gemini,
    "groq": _chat_groq,
    "cerebras": _chat_cerebras,
    "mistral": _chat_mistral,
    "cohere": _chat_cohere,
    "nvidia": _chat_nvidia,
    "huggingface": _chat_huggingface,
    "github": _chat_github,
    "cloudflare": chat_cloudflare,
    "spark": chat_spark,
    "zhipuai": _chat_zhipuai,
    "deepseek": _chat_deepseek,
    "openai": _chat_openai,
    "anthropic": _chat_anthropic,
}
for _pname in OPENAI_COMPAT_PROVIDERS:
    _DISPATCH[_pname] = partial(chat_openai_compat, _pname)


def chat_once(provider: str, messages: List[Dict[str, Any]], model: Optional[str] = None,
              source: str = "") -> Dict[str, Any]:
    """
    Send one chat turn to ``provider`` and return the unified contract.

    Args:
        provider: one of 'openrouter' | 'gemini' | 'deepseek'.
        messages: list of {role, content} dicts (system/user/assistant).
        model:    optional model id; falls back to the provider default.
        source:   task label recorded in the shared usage log ("chat", "compose"…).
    """
    provider = (provider or "").strip().lower()
    # Pass the caller's model through as-is (None = let the handler resolve a
    # live/default model). out["model"] is filled in by the handler.
    requested_model = (model or "").strip() or None
    out = _result(provider, requested_model or "")

    handler = _DISPATCH.get(provider)
    if handler is None:
        out["error"] = f"Unknown provider: {provider!r}"
        return out

    msgs = _normalize_messages(messages)
    if not msgs:
        out["error"] = "No message provided"
        return out

    key = first_secret(provider)
    if not is_configured(provider):
        out["error"] = "No API key configured"
        return out

    use_model = requested_model or default_model(provider)
    rate = check_rate_limit(provider, use_model)
    if not rate.allowed:
        out["error"] = rate.message
        out["retry_after_s"] = rate.retry_after_s
        out["model"] = use_model
        out["nickname"] = chat_nickname(provider, use_model)
        record_usage("text", provider, use_model, False, None, source, rate.message)
        return out

    start = time.time()
    try:
        handler(msgs, requested_model, key, out)
    except Exception as e:  # noqa: BLE001 — surface any SDK failure to the UI
        out["error"] = str(e)
        ColorPrint.yellow(f"[ai_chat] {provider} chat failed: {e}")
    if not out["model"]:
        out["model"] = default_model(provider)
    out["nickname"] = chat_nickname(provider, out["model"])
    out["latency_ms"] = round((time.time() - start) * 1000, 1)
    if out.get("success"):
        record_request(provider)
    record_usage("text", provider, out["model"], bool(out.get("success")),
                 out["latency_ms"], source, out.get("error"))
    return out


__all__ = ["chat_once"]
